### PR TITLE
[wip] add precision to retriever mvp

### PR DESCRIPTION
**Pull request by bbqiu**

### CODE REVIEW

File: mlflow/metrics/metric_definitions.py [hunk 1]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
 
 
-def _validate_text_data(data, metric_name, column_name):
-    """Validates that the data is text and is non-empty"""
-    if len(data) == 0:
-        return False
+def _is_text_data_invalid(data, metric_name, column_name):
+    """Validates that data list a list of strs and is non-empty"""
+    if data is None or len(data) == 0:
+        return True
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:22]: The comment says "Validates that data list a list of strs and is non-empty". Should it be `False` here?

File: mlflow/metrics/metric_definitions.py [hunk 2]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
     }
 
 
-def _validate_text_data(data, metric_name, column_name):
-    """Validates that the data is text and is non-empty"""
-    if len(data) == 0:
-        return False
+def _is_text_data_invalid(data, metric_name, column_name):
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:19]: It seems a bit confusing to me to assert "invalid" than the original "valid"...

File: mlflow/metrics/metric_definitions.py [hunk 3]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
+    """Validates that data list a list of strs and is non-empty"""
+    if data is None or len(data) == 0:
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:21]: We can keep this change on L20-21.

File: mlflow/metrics/metric_definitions.py [hunk 4]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
+        return True
 
     for row, line in enumerate(data):
         if not isinstance(line, str):
             _logger.warning(
                 f"Cannot calculate {metric_name} for non-string inputs. "
-                + f"Non-string found for {column_name} on row {row}. skipping metric logging."
+                f"Non-string found for {column_name} on row {row}. skipping metric logging."
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:28]: Can you revert all the reformatting changes to the unrelated parts in this PR and put them to a separate PR?
In general, it makes the current PR easier to review.

File: mlflow/metrics/metric_definitions.py [hunk 5]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
-            return False
+            return True
 
-    return True
+    return False
+
+
+# used for retriever evaluation with mutliple doc ground truths
+def _are_docs_invalid(data, metric_name, column_name):
+    """Validates that the data is a list of a tuple of strings and is non-empty"""
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:37]:
```suggestion
def _validate_text_tuple_data(data, metric_name, column_name):
    """Validates that the data is a list of tuples of strings and is non-empty"""
```

File: mlflow/metrics/metric_definitions.py [hunk 6]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
+    if data is None or len(data) == 0:
+        return True
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:39]:
```suggestion
        return False
```

File: mlflow/metrics/metric_definitions.py [hunk 7]
```diff
@@ -16,20 +16,37 @@ def standard_aggregations(scores):
...
+
+    for row, docs in enumerate(data):
+        if not isinstance(docs, tuple) or not all(isinstance(doc, str) for doc in docs):
+            _logger.warning(
+                f"Cannot calculate {metric_name} for non-string inputs. "
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:44]:
```suggestion
                f"Cannot calculate {metric_name} for non-Tuple[str] inputs. "
```

File: mlflow/metrics/metric_definitions.py [hunk 8]
```diff
@@ -54,7 +71,7 @@ def _token_count_eval_fn(predictions, targets, metrics):
 
 
 def _toxicity_eval_fn(predictions, targets, metrics):
-    if not _validate_text_data(predictions, "toxicity", "predictions"):
+    if _is_text_data_invalid(predictions, "toxicity", "predictions"):
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:74]: Can you move unrelated changes to a separate PR? Same below.

File: mlflow/metrics/metric_definitions.py [hunk 9]
```diff
@@ -371,3 +361,19 @@ def _f1_score_eval_fn(
...
+def _precision_at_k_eval_fn(predictions, targets, k, metrics, sample_weight=None):
+    if _are_docs_invalid(predictions, "precision_at_k", "predictions") or _are_docs_invalid(
+        targets, "precision_at_k", "targets"
+    ):
+        return
+
+    scores = []
+    for i in range(len(predictions)):
+        # only include the top k retrieved chunks
+        ground_truth, retrieved = set(predictions[i]), targets[i][: k[i]]
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:375]: 1. What if k[i] > len(predictions[i])?
2. 
```suggestion
       retrieved, ground_truth = set(predictions[i]), targets[i][: k[i]]
```

File: mlflow/metrics/metric_definitions.py [hunk 10]
```diff
@@ -371,3 +361,19 @@ def _f1_score_eval_fn(
...
+        relevant_doc_count = sum(1 for doc in retrieved if doc in ground_truth)
+        scores.append(relevant_doc_count / len(retrieved))
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:377]: It's a normal case where `len(retrieved)==0`, in this case, let's return precision=1 for such rows.

File: mlflow/metrics/metric_definitions.py [hunk 11]
```diff
@@ -371,3 +361,19 @@ def _f1_score_eval_fn(
             sample_weight=sample_weight,
         )
         return MetricValue(aggregate_results={"f1_score": f1})
+
+
+def _precision_at_k_eval_fn(predictions, targets, k, metrics, sample_weight=None):
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:366]: Is it possible to pass `k` as a scalar value? Can we validate k is int and `k>0`?

File: tests/evaluate/test_default_evaluator.py [hunk 12]
```diff
@@ -2927,3 +2927,75 @@ def test_evaluate_with_latency():
         "token_count",
     }
     assert all(isinstance(grade, float) for grade in logged_data["latency"])
+
+
+def test_evaluate_retriever():
```
Comment (liangz1) [inline tests/evaluate/test_default_evaluator.py:2932]: Can we test the case where the retrieved docs is an empty tuple?

File: mlflow/metrics/metric_definitions.py [hunk 13]
```diff
@@ -32,6 +32,22 @@ def _validate_text_data(data, metric_name, column_name):
...
+def _validate_text_tuple_data(data, metric_name, column_name):
+    """Validates that the data is a list of a tuple of strings and is non-empty"""
+    if data is None or len(data) == 0:
+        return False
+
+    for row, docs in enumerate(data):
+        if not isinstance(docs, tuple) or not all(isinstance(doc, str) for doc in docs):
+            _logger.warning(
+                f"Cannot calculate {metric_name} for non-Tuple[str] inputs. "
+                f"Non-string found for {column_name} on row {row}. skipping metric logging."
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:44]:
```suggestion
                f"Non-Tuple[str] found for {column_name} on row {row}. skipping metric logging."
```

Comment (liangz1) [reply mlflow/metrics/metric_definitions.py:44]: Can we print `docs` in the error message?

File: mlflow/metrics/metric_definitions.py (hunk 11 above)
Comment (bbqiu) [reply mlflow/metrics/metric_definitions.py:380]: had to rebase to include sunish's latest PR, but it works now!

File: mlflow/metrics/metric_definitions.py [hunk 14]
```diff
@@ -33,6 +33,28 @@ def _validate_text_data(data, metric_name, column_name):
...
 
+# because single entry tuples get unpacked, put them back into a tuple if the entry is a string
+def _validate_and_fix_text_tuple_data(data, metric_name, column_name):
+    """Validates that the data is a list of a tuple of strings and is non-empty"""
+    if data is None or len(data) == 0:
+        return False
+
+    for row, tup in enumerate(data):
+        if not isinstance(tup, tuple) or not all(isinstance(val, str) for val in tup):
+            if isinstance(tup, str):
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:44]: Can we move the comment on L36 to here?

File: mlflow/metrics/metric_definitions.py [hunk 15]
```diff
@@ -33,6 +33,28 @@ def _validate_text_data(data, metric_name, column_name):
...
+                data[row] = (tup,)
+                tup = data[row]
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:46]: Do we need L46?

File: mlflow/metrics/metric_definitions.py [hunk 16]
```diff
@@ -359,3 +381,25 @@ def _f1_score_eval_fn(
...
         )
         return MetricValue(aggregate_results={"f1_score": f1})
+
+
+def _precision_at_k_eval_fn(predictions, targets, k, metrics, sample_weight=None):
+    if (
+        not _validate_and_fix_text_tuple_data(predictions, "precision_at_k", "predictions")
+        or not _validate_and_fix_text_tuple_data(targets, "precision_at_k", "targets")
+        or not isinstance(k, int)
+        and k > 0
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:391]:
```suggestion
        or not (isinstance(k, int) and k > 0)
```

Comment (liangz1) [reply mlflow/metrics/metric_definitions.py:391]: Shall we print a warning message about why k fails the validation? Currently if k fails the validation, it just fails silently and the user does not know what's going wrong.

File: tests/evaluate/test_default_evaluator.py [hunk 17]
```diff
@@ -3184,3 +3184,95 @@ def test_evaluate_custom_metrics_string_values():
...
+    with mlflow.start_run() as run:
+        mlflow.evaluate(
+            model=fn3,
+            data=X,
+            targets="ground_truth",
+            model_type="retriever",
+            evaluator_config={
+                "default": {
+                    "k": 3,
+                }
```
Comment (liangz1) [inline tests/evaluate/test_default_evaluator.py:3270]: Can we add a test case to test the error handling for `evaluator_config={}`?

File: mlflow/metrics/metric_definitions.py [hunk 18]
```diff
@@ -359,3 +381,25 @@ def _f1_score_eval_fn(
             sample_weight=sample_weight,
         )
         return MetricValue(aggregate_results={"f1_score": f1})
+
+
+def _precision_at_k_eval_fn(predictions, targets, k, metrics, sample_weight=None):
```
Comment (liangz1) [inline mlflow/metrics/metric_definitions.py:386]: Can we remove the function define on L279?

File: mlflow/metrics/__init__.py [hunk 19]
```diff
@@ -265,7 +265,10 @@ def rougeLsum() -> EvaluationMetric:
 
 
 precision_at_k = make_metric(
-    eval_fn=_precision_at_k_eval_fn, greater_is_better=True, name="precision_at_k", version="v1"
+    eval_fn=_precision_at_k_eval_fn,
+    greater_is_better=True,
+    name="precision_at_k",
+    version="v1",
```
Comment (liangz1) [inline mlflow/metrics/__init__.py:271]: Can we update the docstring on L273-L281?

File: tests/evaluate/test_default_evaluator.py [hunk 20]
```diff
@@ -3184,3 +3184,95 @@ def test_evaluate_custom_metrics_string_values():
...
+            evaluator_config={
+                "k": 3,
+            },
+        )
+    run = mlflow.get_run(run.info.run_id)
+    assert run.data.metrics == {
+        "precision_at_k/v1/mean": 2 / 3,
+        "precision_at_k/v1/variance": 0,
+        "precision_at_k/v1/p90": 2 / 3,
+    }
```
Comment (liangz1) [inline tests/evaluate/test_default_evaluator.py:3211]: Can we also assert that the per-row metrics are logged in the eval_results_table.json as an artifact? Checking the existence of this file for one test case is sufficient. We just want to make sure it presents for the `precision_at_k` metric.

**Review by liangz1 — APPROVED** [review]
Thanks @bbqiu ! Feel free to merge this to the feature branch after addressing the comments.
